index.html

```diff
@@ -1,453 +1,585 @@
       <p>
-        This specification defines a cryptographic suite for the purpose of
-        creating, verifying proofs for EIP712 signatures in conformance
-        with the Linked Data Proofs [[LD-PROOFS]] specification.
-      </p>
-      <p>
-        In general the suites uses the JCS Algorithm
-        [[JCS]] to transform an input document into its
-        canonical form. The cannonical representation is then provided 
-        to the EIP712 signature function.
+        If <code>TypedData</code>'s <code>types</code> object is not provided to the signature suite, the suite MUST
+        generate the JSON object by
+        inferring types from the input document, and optionally a provided `primaryType`.
       </p>
 
       <p>
-        A signature scheme consists of a hashing algorithm and a signing algorithm.
-        The signing algorithm of choice in Ethereum is secp256k1. The hashing algorithm of
-        choice is keccak256.
+        In case of ambiguous types, the algorithm SHOULD defer to using the most liberal option. For example, a number
+        should be inferred as the <code>uint256</code> type even if that specific number can fit in a <code>uint8</code>
+        type.
       </p>
 
       <p>
-        EIP712 defines a standard API for Web3 Provider (e.g., wallets) to generate signatures
-        over human-readable data where the signature can be verified either by a Smart Contract
-        on the Ethereum Blockchain, or completely offchain. 
-      </p>
+        The <a>types generation algorithm</a> is defined as follows:
 
-      <p>        
-        The rational is to use existing Web3 Providers and their secure key management system
-        to produce signatures that are compliant with the JSON-LD and more specifically, the
-        Linked Data Signatures (LDS) data model.
       </p>
 
-      <p>        
-        Since the EIP712 signature function relies
-        on JSON schemas, implementers need to ensure that the linked data document matches the
-        EIP712 JSON schema that will be provided to the EIP712 signature function.
-      </p>        
+      <ol class="algorithm">
+        <li>Creates a mapping <code>output</code> from <code>string</code> to <code>TypedDataField[]</code> types, where
+          <code>TypedDataField</code> is an object consisting of two string properties - <code>name</code> and
+          <code>type</code>
+        </li>
+        <li>Creates an empty array <code>types</code> of <code>TypedDataField</code> to collect all the fields</li>
+        <li>Canonicalizes the input document using the <a href="#dfn-canonicalization-algorithm">canonicalization
+            algorithm</a></li>
+        <li>If `primaryType` is not provided, set `primaryType = "Document"` else use the provided value.</li>
+        <li>For each property in the canonicalized input document, the algorithm checks the type of the value specific
+          to the
+          implementation language.</li>
```

Comment:
As mentioned in https://github.com/w3c-ccg/ethereum-eip712-signature-2021-spec/pull/25#discussion_r697449235 - I still think it is important to say the iteration order of the object properties, so that the type arrays are constructed consistently. Even though JCS has been used on the input, JCS's canonicalization is in its serialization, but we do not use the serialized representation, instead it is effectively immediately de-serialized back to JSON objects, so arguably canonicalization has not taken place. Iteration order of JavaScript object properties is not necessarily stable: https://stackoverflow.com/questions/30076219/does-es6-introduce-a-well-defined-order-of-enumeration-for-object-properties
Here's some possible text to address this:
```suggestion
          implementation language. Iterate through the properties in lexicographic order of property name according to <a href="https://datatracker.ietf.org/doc/html/rfc8785#section-3.2.3">RFC 8785 Section 3.2.3</a>.</li>
```
Alternatively, the entries in `types` could be sorted by name at the end of the algorithm.

Reply:
We can add that and essentially remove JCS from the algorithm since as you mentioned we are not seeing any real benefits of it. This is news to me and I wasn't completely aware that `JSON.parse(...)` is not stable, but looking into it now I agree with you.

We can just sort the keys recursively and generate `types` based on that - which is essentially what JCS also does except it sorts the keys and builds a string representation recursively. How do you feel about removing the language around JCS entirely?

Reply:
I agree and feel that would be appropriate.

The entry in the Suite Definition table for *canonicalization algorithm* could then be changed to refer to this specification itself or a subsection where it is specifically defined.

When specifying sorting (ordering) of strings (property names / object keys), it should be specified how to sort them. Referencing RFC 8785 Section 3.2.3 would accomplish this I think. RFC 8785 Section 3.2.3 specifies to compare property name strings as arrays of UTF-16 code units, and states that this "maps directly to the string type in ECMAScript". I think that means that Browser and Node.js implementations can use `Array.prototype.sort` on arrays of such strings.